plugins/hive-plugin/src/main/java/edu/berkeley/ground/plugins/hive/GDatabase.java

```diff
@@ -0,0 +1,281 @@
+        } catch (GroundException e) {
+            LOG.debug("Not found - database table edge structure: " + nodeVersion.getNodeId());
+            throw e;
+        }
+    }
+
+    public Database fromJSON(String json) {
+        Gson gson = new Gson();
+        return (Database) gson.fromJson(json, Database.class);
+    }
+
+    public String toJSON(Database db) {
+        Gson gson = new Gson();
+        return gson.toJson(db);
+    }
+
+    public Database getDatabase(String dbName) throws NoSuchObjectException {
+        try {
+            List<String> versions = ground.getNodeFactory().getLeaves(dbName);
+            if (versions.isEmpty()) {
+                throw new GroundException("Database node not found: " + dbName);
+            }
+
+            NodeVersion latestVersion = ground.getNodeVersionFactory().retrieveFromDatabase(versions.get(0));
+            Map<String, Tag> dbTag = latestVersion.getTags();
+
+            return this.fromJSON((String) dbTag.get(dbName).getValue());
+        } catch (GroundException e) {
+            throw new NoSuchObjectException(e.getMessage());
+        }
+    }
+
+    public NodeVersion createDatabase(Database db) throws InvalidObjectException, MetaException {
+        if (db == null) {
+            throw new InvalidObjectException("Database object passed is null");
+        }
+        try {
+            String dbName = db.getName();
+            Node dbNode = this.getNode(dbName);
+            Structure dbStruct = this.getNodeStructure(dbName);
+
+            Tag dbTag = new Tag("1.0.0", dbName, toJSON(db), GroundType.STRING);
+
+            Map<String, GroundType> structVersionAttribs = new HashMap<>();
+            structVersionAttribs.put(dbName, GroundType.STRING);
+
+            StructureVersion sv = ground.getStructureVersionFactory().create(dbStruct.getId(), structVersionAttribs, EMPTY_PARENT_LIST);
+
+            String reference = db.getLocationUri();
+            HashMap<String, Tag> tags = new HashMap<>();
+            tags.put(dbName, dbTag);
+
+            Map<String, String> dbParamMap = db.getParameters();
+            if (dbParamMap == null) {
+                dbParamMap = new HashMap<String, String>();
+            }
+            Map<String, String> parameters = dbParamMap;
+
+            List<String> parent = new ArrayList<String>();
+            List<String> versions = ground.getNodeFactory().getLeaves(dbName);
+            if (!versions.isEmpty()) {
+                parent.add(versions.get(0));
+            }
+
+            NodeVersion dbNodeVersion = ground.getNodeVersionFactory().create(tags, sv.getId(), reference, parameters, dbNode.getId(), parent);
+
+            return dbNodeVersion;
+        } catch (GroundException e) {
+            LOG.error("Failure to create a database node: " + db.getName());
+            throw new MetaException(e.getMessage());
+        }
+    }
+
+    // Table related functions
```

Comment:
The code walks down the hierarchy from metastore to database, then table and finally to parittion for all calls. This is so that changes made at the leaf can have the corresponding effect up the tree ultimately changing the metastore node at the top.